tests/test_plan.py

```diff
@@ -498,3 +498,112 @@ def __call__(self,
             'b3': Annotated[B, 43],
         }),
     ]
+
+
+class Item:
+    def __init__(self, field: int = 42):
+        self.field = field
+
+
+def get_item_factory(page_cls: Callable) -> Callable:
+    def item_factory(page: page_cls) -> Item:  # type: ignore[valid-type]
+        return page.to_item()  # type: ignore[attr-defined]
+
+    return item_factory
+
+
+def get_custom_builders(page_cls: Callable) -> Dict[Callable, Callable]:
+    return {
+        Item: get_item_factory(page_cls),
+    }
+
+
+def test_plan_custom_builder():
+    # Item is built from Page
+
+    class Page:
+        def __init__(self, b: B):
+            pass
+
+        def to_item(self):
+            return Item()
+
+    def fn(item: Item) -> int:
+        return item.field + 1
+
+    custom_builders = get_custom_builders(Page)
+    plan = andi.plan(
+        fn,
+        is_injectable={B, Page},
+        externally_provided={Item},
+        custom_builder_fn=custom_builders.get
+    )
+    assert plan == [
+        (B, {}),
+        (Page, {"b": B}),
+        (CustomBuilder(Item, custom_builders[Item]), {"page": Page}),
+        (fn, {"item": Item})
+    ]
+    instances = build(plan)
+    assert set(instances.keys()) == {B, Page, Item, fn}
+    assert instances[fn] == 43
+
+
+def test_plan_custom_builder_modify_item():
+    # Item is built directly (externally provided)
+    # Page is built from that Item
+    # The final Item is built from Page
+
+    class Page:
+        def __init__(self, b: B, item: Item):
+            self.item = item
+
+        def to_item(self):
+            return Item(self.item.field * 2)
+
+    def fn(item: Item) -> int:
+        return item.field + 1
+
+    custom_builders = get_custom_builders(Page)
+    plan = andi.plan(
+        fn,
+        is_injectable={B, Page},
+        externally_provided={Item},
+        custom_builder_fn=custom_builders.get
+    )
+    assert plan == [
+        (B, {}),
+        (Item, {}),
```

Comment:
This randomly fails on 3.5 because these two entries are sometimes swapped.

Reply:
Looks like the underlying issue is the iteration order of various dicts representing args, which on 3.5 is random between runs.